### PR TITLE
Read data from source file in Portal and write to MIG drop bucket

### PR DESCRIPTION
**Pull request by atrout**
### Description and Motivation
Code to extract data from Phoenix, optionally create a new MIG dataset, and write data to MIG destination bucket.
I think this represents the basic end-to-end functionality.

I'm going to extract the BigQuery functions into a separate helper file / class in a follow-up PR.

### Changes
Added get_schema, create_dataset, get_source_data, and write_data_to_file functions

### Testing
Will test manually when merged. I have tested all the composite chunks of functionality outside of Script Runner.

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -1,25 +1,130 @@
...
+    # Log results
 
 # Check if dataset of specified name already exists
-# If dataset doesn't exist,
-    # get schema of source table and create MIG dataset
-# Get data from source dataset
-# Get signed uploadurl from MIG
-# Upload the data for the dataset in MIG to presigned url
-# Log results
+installations = dx.get_installations()
```
Comment (forklady42) [inline run.py]: Would prefer having these calls in a `__main__` block, though not a blocker.

**Review by forklady42 — APPROVED** [review]

**Comment by chrisgoddard** [comment]
Few things:

## Globals

There are quite a few globals being used which can make things harder to debug.

Suggest bundling all of the main logic into a "main" function at the bottom and then wrapping a call to it like:

```python

if __name__ = '__main__':
    main()
    
 ```
 
 
Instantiate and dependencies within the main function and pass them to the functions that need them, e.g.

Also, all logic should be orchestrated from the main function so instead of having `write_data_to_file` running the `get_source_data` function, get_source_data is run in `main` and then passes the data through to the `write_data_to_file` function

```python

def get_source_data(client: bigquery.Client) -> list:
    """
    Get data from Portal source dataset
    """
    sql = f"""
        SELECT TO_JSON_STRING(t) json
        FROM (
            SELECT *
            FROM `{dataset_id}.{table_name}`
        ) t
    """
    query_job = client.query(sql)
    rows = query_job.result()
    data = []
    for row in rows:
        print(row.values()[0])
        data.append(row.values()[0])
    return data

...

def main():
   ...
   
   client = bigquery.Client(credentials=credentials, project=project)
   
   source_data = get_source_data(client)
   
   write_data_to_file(destination_dataset, source_data)
   
   ...
   
   
```

## Installations

I *think* most clients will only have a single installation but I think we should have the option to define an installation ID and at least check it, i.e.

```python

target_installation_id = os.environ.get('TARGET_INSTALL_ID')

target_install = None

installations = dx.get_installations()

if len(installations) == 0:
    raise Exception("No valid installation found")
elif len(installations) == 1:
    target_install = installations[0]
    if target_installation_id:
        if target_install.movement_app_installation_id != int(target_installation_id):
            raise Exception(f"Installation {target_installation_id} requested but only {target_install.movement_app_installation_id} was found")
 elif len(installations) > 1 and target_installation_id is None:
     raise Exception("More than one installation available and not TARGET_INSTALL_ID specifieid")
 else:
     for install in installations:
         if install.movement_app_installation_id == int(target_installation_id):
             target_install = install
             break
     if target_install is None:
         raise Exception("Installation {target_installation_id} not found")
    


```

I would raise Exceptions if there are any conditions that occur that would prevent the script from running as I believe these will be logged in the failure notification and be accessible to the user (we should check this).


## Runtime Arguments

I think from looking at the documentation that any runtime variables (i.e. the table name) are going to be passed to the script by injecting the environment variables into the startup command - e.g. `python run.py --variable $ENV_VARIABLE`

If that's the case would it make more sense to just use environment variables for everything? The start of `main` could check that all the environment variables we need exist and raise an exception if there's anything missing.

## Google Service Account

We should double check but I think the Google Service Account JSON credential will need to be passed via an environment variable too - so we should make sure we're doing the Google Auth using a credential string rather the APPLICATION_DEFAULT which I think is what you're using now.

**Comment by atrout** [comment]
> ## Runtime Arguments
> I think from looking at the documentation that any runtime variables (i.e. the table name) are going to be passed to the script by injecting the environment variables into the startup command - e.g. `python run.py --variable $ENV_VARIABLE`
> 
> If that's the case would it make more sense to just use environment variables for everything? The start of `main` could check that all the environment variables we need exist and raise an exception if there's anything missing.

Yes, the parameters are passed to the script in the startup command (though I was unable to find this documented anywhere -- I had to watch a recorded zoom presentation to find that, so I'd be interested in knowing where you saw that). I think there's a chance that at least the table name might be dynamic / could be determined or overwritten by previous steps in the job, so it might be beneficial to have those be parameters rather than environment variables. I'm going to leave them as parameters for now but we could easily change them to environment variables.

**Comment by atrout** [comment]
> I _think_ most clients will only have a single installation but I think we should have the option to define an installation ID and at least check it, i.e.

Thanks for this info. I'm going to backlog a ticket to handle this later.

**Comment by atrout** [comment]
> We should double check but I think the Google Service Account JSON credential will need to be passed via an environment variable too - so we should make sure we're doing the Google Auth using a credential string rather the APPLICATION_DEFAULT which I think is what you're using now.

Can you say a little more about this? What is it that requires the google credentials? We need google creds on the Portal side for reading from BigQuery but I was kind of hoping those creds would automatically be included by ScriptRunner (part of what I was hoping to test when this PR is merged).

**Comment by chrisgoddard** [comment]
> > We should double check but I think the Google Service Account JSON credential will need to be passed via an environment variable too - so we should make sure we're doing the Google Auth using a credential string rather the APPLICATION_DEFAULT which I think is what you're using now.
> 
> Can you say a little more about this? What is it that requires the google credentials? We need google creds on the Portal side for reading from BigQuery but I was kind of hoping those creds would automatically be included by ScriptRunner (part of what I was hoping to test when this PR is merged).

So I reread the documentation and I *think* the script runs in a context where the service account authorization is already present (likely using APPLICATION_DEFAULT_CREDENTIALS) - but otherwise we'd have to pass the credentials.json file as a string via environment variable. It's not difficult but ideally we don't have to.

I think it's ok to leave for now but if we run into permission issues when testing we should check this off as a variable

**Pull request merged**